storage/rocksdb/ha_rocksdb.cc

```diff
@@ -5794,55 +5840,179 @@ class Rdb_writebatch_impl : public Rdb_transaction {
+  } else {
+    tail = tx->prev;
+  }
+}
+
+void Rdb_transaction_list::fixup_newest_snapshot_after_erase(
+    const Rdb_transaction *erased_tx) {
+  if (unlikely(erased_tx == newest_tx_with_snapshot)) {
+    assert(erased_tx->get_earliest_snapshot_ts() != 0);
+    newest_tx_with_snapshot = erased_tx->prev;
+    assert(newest_tx_with_snapshot == nullptr ||
+           newest_tx_with_snapshot->get_earliest_snapshot_ts() != 0);
+  }
+#ifndef NDEBUG
+  if (newest_tx_with_snapshot != nullptr) {
+    check_node(newest_tx_with_snapshot,
+               newest_tx_with_snapshot->get_earliest_snapshot_ts());
   }
+#endif
+}
 
+void Rdb_transaction_list::erase(Rdb_transaction *tx) {
+  RDB_MUTEX_LOCK_CHECK(mutex);
+  do_erase(tx);
+  fixup_newest_snapshot_after_erase(tx);
+  RDB_MUTEX_UNLOCK_CHECK(mutex);
+}
+
+void Rdb_transaction_list::reinsert_with_snapshot(Rdb_transaction *tx,
+                                                  std::int64_t ts) {
+  assert(ts != 0);
+  assert(tx->get_earliest_snapshot_ts() == 0);
+
+  RDB_MUTEX_LOCK_CHECK(mutex);
+
+  assert(newest_tx_with_snapshot != tx);
+  assert(newest_tx_with_snapshot == nullptr ||
+         newest_tx_with_snapshot->get_earliest_snapshot_ts() != 0);
+
+  do_erase(tx);
+
+  tx->m_earliest_snapshot_ts = ts;
+
+  auto *preceding_tx = newest_tx_with_snapshot;
+  // In regular thread scheduling, the threads are expected to acquire snapshots
+  // and get reinserted in the same order, in which case the list will not be
+  // traversed.
+  while (unlikely(preceding_tx != nullptr &&
+                  preceding_tx->get_earliest_snapshot_ts() >
+                      tx->get_earliest_snapshot_ts())) {
+#ifndef NDEBUG
+    check_node(preceding_tx, preceding_tx->get_earliest_snapshot_ts());
+#endif
+    preceding_tx = preceding_tx->prev;
+  }
+
+  if (unlikely(preceding_tx == nullptr)) {
+    tx->prev = nullptr;
+    if (likely(head != nullptr)) head->prev = tx;
+    tx->next = head;
+    head = tx;
+    if (unlikely(tail == nullptr)) tail = tx;
+  } else {
+    tx->next = preceding_tx->next;
+    tx->prev = preceding_tx;
+    if (likely(preceding_tx->next != nullptr)) {
+      preceding_tx->next->prev = tx;
+    } else {
+      tail = tx;
+    }
+    preceding_tx->next = tx;
+  }
+
+  if (preceding_tx == newest_tx_with_snapshot) {
+    newest_tx_with_snapshot = tx;
+  }
+
+#ifndef NDEBUG
+  check_node(tx, ts);
+  check_node(head, head->get_earliest_snapshot_ts());
+  check_node(newest_tx_with_snapshot,
+             newest_tx_with_snapshot->get_earliest_snapshot_ts());
+  check_node(tail, tail->get_earliest_snapshot_ts());
+  assert(oldest_timestamp_locked() <= ts);
+#endif
+
+  RDB_MUTEX_UNLOCK_CHECK(mutex);
+}
+
+void Rdb_transaction_list::reinsert_without_snapshot(Rdb_transaction *tx) {
+  if (tx->get_earliest_snapshot_ts() == 0) return;
+
+  RDB_MUTEX_LOCK_CHECK(mutex);
+  do_erase(tx);
+  fixup_newest_snapshot_after_erase(tx);
+  tx->m_earliest_snapshot_ts = 0;
+  tx->next = nullptr;
```

Comment:
why only clean tx->next? how about tx->prev?

Reply:
`do_insert_without_snapshot` handles `tx->prev`. It does not handle `tx->next` because the other caller sets it outside the mutex critical section, to have it as short as possible.